File: ios/Framework/Core/Models/Comment.swift
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */
import Foundation


@objc public class Comment: NSObject, NSCoding {

	public static func plainBodyToHtml(plainBody: String) -> String {
		return plainBody
			.stringByReplacingOccurrencesOfString("<", withString: "&lt;")
			.stringByReplacingOccurrencesOfString(">", withString: "&gt;")
			.characters
			.split("\n")
			.map(String.init)
			.map { "<p>\($0)</p>" }
			.joinWithSeparator("")
	}

	private let AllowedTags = ["strong", "i", "b", "a"]

	public let attributes: [String:AnyObject]

	public var originalBody: String {
		return attributes["body"]!.description
			.stringByReplacingOccurrencesOfString("\n ", withString: "")
			.stringByReplacingOccurrencesOfString("\n", withString: "")
			.stringByReplacingOccurrencesOfString("</p><p>", withString: "\n")
			.stringByReplacingOccurrencesOfString("<p>", withString: "")
			.stringByReplacingOccurrencesOfString("</p>", withString: "")
	}

	public var plainBody: String {
		return originalBody
			.stringByReplacingOccurrencesOfString(
				"<[^>]+>",
				withString: "",
				options: .RegularExpressionSearch,
				range: nil)
			.stringByReplacingOccurrencesOfString("&lt;", withString: "<")
			.stringByReplacingOccurrencesOfString("&gt;", withString: ">")
	}

	public var htmlBody: String {
		let closeTags = AllowedTags.map { "/\($0)" }
		let allTags = closeTags + AllowedTags

		return originalBody
			.stringByReplacingOccurrencesOfString(
				"(?i)<(?!\(allTags.joinWithSeparator("|"))).*?>",
				withString: "",
				options: .RegularExpressionSearch,
				range: nil)
			.stringByReplacingOccurrencesOfString("\n", withString: "</br>")
	}

	public var isStyled: Bool {
		return originalBody.containsString("<")
	}

	public var commentId: Int64 {
		return (attributes["commentId"]! as! NSNumber).longLongValue
	}

	public var userName: String {
		return attributes["userName"]!.description
	}

	public var userId: Int64 {
		return (attributes["userId"]! as! NSNumber).longLongValue
	}

	public var createDate: NSDate {
		let milliseconds = (attributes["createDate"]! as! NSNumber).doubleValue
		return NSDate(millisecondsSince1970: milliseconds)
	}

	public var modifiedDate: NSDate {
		let milliseconds = (attributes["modifiedDate"]! as! NSNumber).doubleValue
		return NSDate(millisecondsSince1970: milliseconds)
	}

	public var canDelete: Bool {
		return attributes["deletePermission"] as? Bool ?? false
	}

	public var canEdit: Bool {
		return attributes["updatePermission"] as? Bool ?? false
	}

	public func encodeWithCoder(aCoder: NSCoder) {
		aCoder.encodeObject(self.attributes, forKey:"comment-attrs")
	}

	//MARK: Initializers

	public init(attributes: [String:AnyObject]) {
		self.attributes = attributes

		super.init()
	}

	public required init?(coder aDecoder: NSCoder) {
		self.attributes = aDecoder.decodeObjectForKey("comment-attrs") as? [String:AnyObject] ?? [:]

		super.init()
	}
}

File: xamarin/Framework/LiferayScreens.iOS/Models/Comment.cs
using Foundation;
using ObjCRuntime;
using System;

namespace LiferayScreens
{
    // @interface Comment : NSObject <NSCoding>
    [BaseType(typeof(NSObject))]
    [DisableDefaultCtor]
    interface Comment : INSCoding
    {
        // +(NSString * _Nonnull)plainBodyToHtml:(NSString * _Nonnull)plainBody __attribute__((warn_unused_result));
        [Static]
        [Export("plainBodyToHtml:")]
        string PlainBodyToHtml(string plainBody);

        // @property (readonly, copy, nonatomic) NSDictionary<NSString *,id> * _Nonnull attributes;
        [Export("attributes", ArgumentSemantic.Copy)]
        NSDictionary<NSString, NSObject> Attributes { get; }

        // @property (readonly, copy, nonatomic) NSString * _Nonnull originalBody;
        [Export("originalBody")]
        string OriginalBody { get; }

        // @property (readonly, copy, nonatomic) NSString * _Nonnull plainBody;
        [Export("plainBody")]
        string PlainBody { get; }

        // @property (readonly, copy, nonatomic) NSString * _Nonnull htmlBody;
        [Export("htmlBody")]
        string HtmlBody { get; }

        // @property (readonly, nonatomic) BOOL isStyled;
        [Export("isStyled")]
        bool IsStyled { get; }

        // @property (readonly, nonatomic) int64_t commentId;
        [Export("commentId")]
        long CommentId { get; }

        // @property (readonly, copy, nonatomic) NSString * _Nonnull userName;
        [Export("userName")]
        string UserName { get; }

        // @property (readonly, nonatomic) int64_t userId;
        [Export("userId")]
        long UserId { get; }

        // @property (readonly, copy, nonatomic) NSDate * _Nonnull createDate;
        [Export("createDate", ArgumentSemantic.Copy)]
        NSDate CreateDate { get; }

        // @property (readonly, copy, nonatomic) NSDate * _Nonnull modifiedDate;
        [Export("modifiedDate", ArgumentSemantic.Copy)]
        NSDate ModifiedDate { get; }

        // @property (readonly, nonatomic) BOOL canDelete;
        [Export("canDelete")]
        bool CanDelete { get; }

        // @property (readonly, nonatomic) BOOL canEdit;
        [Export("canEdit")]
        bool CanEdit { get; }

        // -(void)encodeWithCoder:(NSCoder * _Nonnull)aCoder;
        [Export("encodeWithCoderComment:")]
        void EncodeWithCoder(NSCoder aCoder);

        // -(instancetype _Nonnull)initWithAttributes:(NSDictionary<NSString *,id> * _Nonnull)attributes __attribute__((objc_designated_initializer));
        [Export("initWithAttributes:")]
        [DesignatedInitializer]
        IntPtr Constructor(NSDictionary<NSString, NSObject> attributes);
    }
}
